### PR TITLE
ci: fix an edge case where multiple versions share the same checksum

### PR DESCRIPTION
**Pull request by alecbcs**
If someone accidentally submits a package with two versions with the same sha256 we might not catch it in the CI/CD checks depending on the version ordering. This is because the dictionary will squash two versions with the same checksum.

This PR adds a check to ensure versions are not squashed and we rightly report if two versions have identical checksums.

Thank you @tldahlgren for finding this!

### CODE REVIEW

**Comment by becker33** [comment]
We discussed passing these as tuples instead of dictionaries. The methods called here are not called anywhere else, so it's fine to change their signatures.

File: lib/spack/spack/test/cmd/ci.py [hunk 1]
```diff
@@ -2202,6 +2215,23 @@ def test_ci_verify_versions_standard_invalid(
         assert "Invalid commit for diff-test@2.1.6" in out
 
 
+def test_ci_verify_versions_standard_duplicates(
+    monkeypatch,
+    mock_packages,
+    mock_git_package_changes,
+    verify_standard_versions_invalid_duplicates,
+    verify_git_versions_invalid,
```
Comment (becker33) [inline lib/spack/spack/test/cmd/ci.py]: Is this really necessary?

Comment (alecbcs) [reply lib/spack/spack/test/cmd/ci.py]: No that's a remnant of copy/past from the method above that uses both git and standard versions.

Good catch, I'll remove and rename the prior method that shouldn't just be `standard_invalid`.

**Comment by alecbcs** [comment]
Discussed with @becker33, turns out tuples were harder to implement in the low level diff parsing method as we rely on set logic to understand changed versions vs added versions. Instead I flipped the dictionary and rewrote the filter function to fix the edge case and better model the versions. Added an additional check to test for regressions in the future.

**Review by becker33 — APPROVED** [review]
